libcst/codemod/commands/convert_type_comments.py

```diff
@@ -56,29 +58,168 @@ def _convert_annotation(raw: str) -> cst.Annotation:
         return cst.Annotation(annotation=cst.SimpleString(f'"{raw}"'))
 
 
+class _FailedToApplyAnnotation:
+    pass
+
+
+class _ArityError(Exception):
+    pass
+
+
+UnpackedTargets: TypeAlias = Union[cst.BaseExpression, "UnpackedTargets"]
+UnpackedAnnotations: TypeAlias = Union[str, List["UnpackedAnnotations"]]
+TargetAnnotationPair: TypeAlias = Tuple[cst.BaseExpression, str]
+
+
+class AnnotationSpreader:
+    """
+    Utilities to help with lining up tuples of types from type comments with
+    the tuples of values with which they should be associated.
+    """
+
+    # TODO: find a 3.6-compatible way to handle TypeAlias
+
+    @staticmethod
+    def _unparse_annotation(
+        expression: ast.expr,
+    ) -> UnpackedAnnotations:
+        if isinstance(expression, ast.Tuple):
+            return [
+                AnnotationSpreader._unparse_annotation(elt) for elt in expression.elts
+            ]
+        else:
+            return ast.unparse(expression)
+
+    @staticmethod
+    def unpack_type_comment(
+        type_comment: str,
+    ) -> UnpackedAnnotations:
+        """
+        Unpack an ast module expression and unparse it into a recursive
+        list of strings matching the tuple structure of the type comment.
+        """
+        # pyre-ignore[16]: the ast module stubs do not have full details
+        annotation_ast = ast.parse(type_comment, "<type_comment>", "eval").body
+        return AnnotationSpreader._unparse_annotation(annotation_ast)
+
+    @staticmethod
+    def unpack_target(
+        target: cst.BaseExpression,
+    ) -> UnpackedTargets:  # pyre-ignore: I'm unsure what the problem is.
```

Comment:
@grievejia any idea why pyre would complain about `UnpackedTargets` not being a valid type here?

The type error I was getting was:
```
libcst/codemod/commands/convert_type_comments.py:108:9 Undefined or invalid type [11]: Annotation `UnpackedTargets` is not defined as a type
```

But
(a) as far as I can tell there's not much difference between `UnpackedAnnotations` and `UnpackedTargets`, so I can't figure out why only one of them complains
(b) even weirder, it doesn't complain on line 127 but that's exactly the same type name

Reply:
> as far as I can tell there's not much difference between UnpackedAnnotations and UnpackedTargets

This definition doesn't make sense to me:
```
UnpackedTargets: TypeAlias = Union[cst.BaseExpression, "UnpackedTargets"]
```
It's like saying `type t = t` in OCaml -- you can't define `t` in terms of just `t` itself without some sort of additional labeling. Otherwise it would be impossible to inductively prove that any object *cannot* have that type. We should probably emit an error there, but iirc that might be tricky at implementation level. 

On the other hand, the definition of `UnpackedTargets` is OK. The self-recursion is prevented by the existence of `List`, and we can inductively say that the inhabitants of this type are integer trees and nothing else.  

Either way, what you are doing relies on Pyre's support for recursive type, which is a non-standard extension and I personally consider that experimental. If possible I would suggest avoiding that and simple do an old-school class hierarchy (e.g. one abstract base `UnpackedAnnotations` and two concrete subtypes `SingleUnpackedAnnotation` and `MultipleUnpackedAnnotations`, where the latter contains attributes of the parent type), which is guaranteed to always work. 

> even weirder, it doesn't complain on line 127 but that's exactly the same type name

Pyre does not guarantee it always emit all errors. We do some filtering to prevent diagnostics from getting too spammy. iirc one of the filtering rule is that if the invalid type error for a given name is reported once, we don't report it elsewhere in the same file.

Reply:
I hesitate to use a class hierarchy because that adds quite a lot of boilerplate - it becomes harder to follow the intent of the code from the types themselves plus I'd need unwrapping and wrapping logic.

I think using recursive types in a codemod ought to be alright, I probably wouldn't use it in the core of LibCST

Reply:
I made a PR to use a class hierarchy: https://github.com/stroxler/LibCST/pull/2

The change seems pretty bad to me, I'd much prefer to stick with recursive types